src/types/acls.rs

```diff
@@ -0,0 +1,387 @@
+use crate::error::Error;
+use crate::parse::{parse_until_done, MapOrNot};
+use crate::types::UnsolicitedResponse;
+use imap_proto::types::AclRight;
+use imap_proto::Response;
+use ouroboros::self_referencing;
+use std::borrow::Cow;
+use std::collections::HashSet;
+use std::fmt::{Display, Formatter};
+use std::sync::mpsc;
+
+/// Specifies how [`Session::set_acl`] should modify an existing permission set.
+#[derive(Debug, Clone, Copy)]
+pub enum AclModifyMode {
+    /// Replace all ACLs on the identifier for the mailbox
+    Replace,
+    /// Add the given ACLs to the identifier for the mailbox
+    Add,
+    /// Remove the given ACLs from the identifier for the mailbox
+    Remove,
+}
+
+/// A set of [`imap_proto::AclRight`]s.
+///
+/// Used as input for [`Session::set_acl`] as output in [`ListRights`], [`MyRights`], and [`AclEntry`]
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
```

Comment:
I don't think this needs `#[non_exhaustive]` since it has private fields already anyway.